### PR TITLE
feat: introduce AbstractCursor and its concrete subclasses

### PR DESCRIPTION
**Pull request by mbroadst**
This change introduces a fundamental redesign of the cursor types in the driver. The first change is to add a new `AbstractCursor`
type, which is only concerned with iterating a cursor (using `getMore`) once it has been initialized. The `_initialize` method must be implemented by subclasses. The concrete subclasses are generally builders for `find` and `aggregate` commands, each providing their own custom initialization method.

NODE-2809

### CODE REVIEW

File: src/cursor/find_cursor.ts
Comment (nbbeeken) [inline src/cursor/find_cursor.ts:165]: Should this check if the cursor has started and throw?

File: src/cursor/find_cursor.ts
Comment (nbbeeken) [inline src/cursor/find_cursor.ts:173]: Since kBuiltOptions is frozen after _initialize this will throw a TypeError if called then. Maybe its overkill should we document that?

**Review by nbbeeken — CHANGES_REQUESTED** [review]
👏  hats off 🎩  to this incredible chunk of cursor work!!

File: package.json
Comment (mbroadst) [reply package.json:29]: it actually shouldn't be, the `cloneDeep` is used as part of command monitoring which is a shipped feature

File: package.json
Comment (mbroadst) [reply package.json:33]: I left the `lodash` dependency in here just to get the changes into the codebase, but will be providing a custom deep clone for the driver in a subsequent PR.

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:72]: I was thinking the name `AbstractCursor` helps convey the context at the point where you're using the class (`FindCursor extends AbstractCursor`). I don't feel very strongly about this, but it's a convention I've used across a number of languages that support abstract classes for many years, force of habit. Do you think calling it plain `Cursor` would help future readers?

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:174]: Good point, though I think it might be something more like:
```
return this[kDocuments].splice(0, Math.min(number ?? this.bufferedCount()));
```

FWIW I really despise the "bufferedX" methods, and think we should defer to some solution like providing a `nextBatch` method.

File: src/cmap/wire_protocol/query.ts
Comment (mbroadst) [reply src/cmap/wire_protocol/query.ts:207]: [$showDiskLoc](https://docs.mongodb.com/manual/reference/operator/meta/showDiskLoc/#metaOp._S_showDiskLoc) was deprecated in 3.2 in favor of `showRecordId`. The method here is building a find query for legacy servers, so we need to "downgrade" the newly named option to send on the query there. I have a follow-up PR that makes building find commands far more understandable

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:382]: I think it was me just pondering whether any builder methods should exist on the `AbstractCursor`, I'll just remove this for now

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:426]: This is coming in a follow-up PR!

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:432]: I think we should provide a better error message. This is also part of the `withReadPreference` PR that I'll pick up from Thomas' work. There we have a common validation method to provide the same type checking and error messages for this method as well as `ReadPreference.fromOptions`

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:505]: Because we very much don't want people to actually override the implementation. A TypeScript marking will hide it from your IDE, but it won't prevent a user from monkey-patching the behavior. A few months ago we discussed an even more aggressive approach of passing the methods in as options to the `AbstractCursor` (a la Node.js streams) but I figured this was slightly more readable.

File: src/cursor/find_cursor.ts
Comment (mbroadst) [reply src/cursor/find_cursor.ts:156]: freezing the options was my cheap way of ensuring the cursor couldn't be built after initialization (it also cuts down a lot on duplicated error messages), but perhaps we should have an explicit boolean and throw a more informative error message. I'll add that now, we can discuss on those changes

File: src/cursor/find_cursor.ts
Comment (mbroadst) [reply src/cursor/find_cursor.ts:148]: yes, I'll add that to the changes for "initialized"

File: package.json
Comment (nbbeeken) [reply package.json:29]: I think we shouldn't need to ship with depended upon types since we don't re-export anything from lodash

File: src/cursor/abstract_cursor.ts
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:73]: I think this is fine, you make good points about the implication of the name

File: src/cursor/abstract_cursor.ts
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: The number param still needs to be made optional, but also looking closer at splice's behavior if TS will allow it you can pass the potentially undefined number directly to splice and it will default to grabbing [0, length of the array]. 

This is moot if you just use the undefined number directly but if not was it intentional to change to using `this[kDocuments].length` as opposed to `this.bufferedCount()`?

Also the documentation still needs a note here that it removes from the buffer.

Or if all this is moot for the pending `nextBatch()` solution, is there a ticket floating around for that?

File: src/cursor/abstract_cursor.ts
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:182]: I think this was resolved by mistake?
I vaguely recall an issue that doesn't make a custom lib possible, but wasn't 100% on that, is it the case?

File: src/cursor/find_cursor.ts
Comment (nbbeeken) [reply src/cursor/find_cursor.ts:157]: I see the `assertUninitialized` method now there, looks good to me, is there some hesitation you have about going this route?

File: src/cursor/abstract_cursor.ts
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:511]: Discussed offline: It's important to us to maintain a barrier for overriding behavior we want to have guaranteed.

File: src/cursor/abstract_cursor.ts
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:483]: We could also consider still making this protected and add a `//@ts-expect-error` comment to the nextDocument function.

File: package.json
Comment (mbroadst) [reply package.json:29]: oh, doh you're right these are the types

File: src/cursor/abstract_cursor.ts
Comment (nbbeeken) [inline src/cursor/abstract_cursor.ts:20]:
```suggestion
const kInitialized = Symbol('initialized');
```

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts]: nice catch, I figured awaiting `next` would do it, but returning that value in an `async` function definitely would have returned the native Promise

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts:480]: Hm, let's hold off on figuring this protected stuff out until a subsequent PR. Maybe we can also try storing `initialize` and `getMore` as properties passed in through options, and stored with symbols to avoid the method showing up in autocomplete

File: src/cursor/find_cursor.ts
Comment (mbroadst) [reply src/cursor/find_cursor.ts:157]: Nope, if it looks good to you then I think we can resolve

File: src/cursor/abstract_cursor.ts
Comment (mbroadst) [reply src/cursor/abstract_cursor.ts]: Fixed this up. We don't currently have a ticket for `nextBatch` explicitly, but [NODE-2860](https://jira.mongodb.org/browse/NODE-2860) suggests it.

File: src/cursor/abstract_cursor.ts
Comment (emadum) [inline src/cursor/abstract_cursor.ts:166]:
```suggestion
    return this[kClosed];
```

I think this should be using the symbol, if we're keeping it.

File: src/cursor/abstract_cursor.ts
Comment (emadum) [inline src/cursor/abstract_cursor.ts:300]: Seems this is only used once and could be moved down into the lower `if` check, maybe with a comment `// needs to emit close`?

File: src/cursor/abstract_cursor.ts
Comment (emadum) [inline src/cursor/abstract_cursor.ts:541]: Do we need to be concerned about what happens if `cursor[kTopology].hasSessionSupport()` is false here?

File: src/cursor/abstract_cursor.ts
Comment (emadum) [inline src/cursor/abstract_cursor.ts:679]: In my opinion this would be a good opportunity to do it the "right" way.

File: test/functional/operation_example.test.js
Comment (emadum) [inline test/functional/operation_example.test.js:4911]:
```suggestion
  // NOTE: unclear whether we should continue to support `rewind` (NODE-2811)
```

File: src/cursor/abstract_cursor.ts
Comment (emadum) [inline src/cursor/abstract_cursor.ts:346]: Should remove the reference to `cursor.rewind()` or add a note about NODE-2811 here.

**Review by emadum — CHANGES_REQUESTED** [review]
A few questions/nits but this looks great. I reviewed the `src` changes closely but only took a surface level look at the test changes; I'll take a closer look at tests in my follow-up review.

File: src/cursor/abstract_cursor.ts
Comment (emadum) [reply src/cursor/abstract_cursor.ts:166]: In terms of *whether* to keep it, I don't see how users get any utility out of it; I've vote to remove it if it doesn't take too much rewriting of tests.

Comment (mbroadst) [reply src/cursor/abstract_cursor.ts]: I updated this to remove the `isClosed` method on `AbstractCursor` and `ChangeStream`, instead depending on the `closed` accessor

File: test/functional/abstract_cursor.test.js
Comment (emadum) [inline test/functional/abstract_cursor.test.js:53]:
```suggestion
      'should send a killCursors command when closed before completely iterated',
```

**Review by emadum — APPROVED** [review]
LGTM! 👏

**Review by nbbeeken — APPROVED** [review]
lgtm!

**Pull request merged**
